proof/access-control/ARM/ArchInterrupt_AC.thy

```diff
@@ -87,17 +87,12 @@ lemma arch_decode_irq_control_invocation_authorised[Interrupt_AC_assms]:
        (args \<noteq> [] \<longrightarrow> (pasSubject aag, Control, pasIRQAbs aag (ucast (args ! 0))) \<in> pasPolicy aag))\<rbrace>
    arch_decode_irq_control_invocation info_label args slot caps
    \<lbrace>\<lambda>x _. arch_authorised_irq_ctl_inv aag x\<rbrace>, -"
-  unfolding decode_irq_control_invocation_def arch_decode_irq_control_invocation_def
+  unfolding decode_irq_control_invocation_def arch_decode_irq_control_invocation_def Let_def
             authorised_irq_ctl_inv_def arch_authorised_irq_ctl_inv_def arch_check_irq_def
   apply (rule hoare_gen_asmE)
-  apply (rule hoare_pre)
-   apply (simp add: Let_def split del: if_split cong: if_cong)
-   apply (wp whenE_throwError_wp hoare_vcg_imp_lift hoare_drop_imps
-          | strengthen  aag_Control_owns_strg
-          | simp add: o_def del: hoare_True_E_R)+
+  apply (wpsimp wp: weak_if_wp)
```

Comment:
could you say something about what `weak_if_wp` is or when to use it? (not in the commits, but I've either not heard of it or completely forgot)

Reply:
```
⟦ ⦃P⦄ f ⦃Q⦄; ⦃P'⦄ f ⦃Q'⦄ ⟧ ⟹ ⦃P and P'⦄ f ⦃λr. if C r then Q r else Q' r⦄
```
In my mind it's for when there's an `if` in the postcondition where you don't need to know the `if` condition and don't want to lift it across the function for whatever reason.

From memory, my cases in this PR were generally `if` conditions on the return value that would have been painful to lift, and where the two branches were either roughly the same or one was a form of `True`. As an alternative I probably could have used the simplifier to handle the `if`, but this approach with `weak_if_wp` was the one I found first.

Reply:
That's awesome, I hope I won't forget it. It feels better than exciting combination of `if_apply_def2` and `hoare_drop_imps`, and should work in more cases. It's interesting that this isn't achieved with a `wp <some_rule>`, but I guess it does need to shuffle the order.